components/camera/rtsp/rtsp.go

```diff
@@ -0,0 +1,281 @@
+	"context"
+	"image"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/aler9/gortsplib/v2"
+	"github.com/aler9/gortsplib/v2/pkg/base"
+	"github.com/aler9/gortsplib/v2/pkg/liberrors"
+	"github.com/aler9/gortsplib/v2/pkg/url"
+	"github.com/edaniels/golog"
+	"github.com/edaniels/gostream"
+	"github.com/pion/rtp"
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+	goutils "go.viam.com/utils"
+
+	"go.viam.com/rdk/components/camera"
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/registry"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/rimage/transform"
+	"go.viam.com/rdk/utils"
+)
+
+var model = resource.NewDefaultModel("rtsp")
+
+func init() {
+	registry.RegisterComponent(camera.Subtype, model, registry.Component{
+		Constructor: func(ctx context.Context, _ registry.Dependencies, cfg config.Component, logger golog.Logger) (interface{}, error) {
+			attrs, ok := cfg.ConvertedAttributes.(*Attrs)
+			if !ok {
+				return nil, utils.NewUnexpectedTypeError(attrs, cfg.ConvertedAttributes)
+			}
+			return NewRTSPCamera(ctx, attrs, logger)
+		},
+	})
+
+	config.RegisterComponentAttributeMapConverter(
+		camera.Subtype,
+		model,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf Attrs
+			attrs, err := config.TransformAttributeMapToStruct(&conf, attributes)
+			if err != nil {
+				return nil, err
+			}
+			result, ok := attrs.(*Attrs)
+			if !ok {
+				return nil, utils.NewUnexpectedTypeError(result, attrs)
+			}
+			return result, nil
+		},
+		&Attrs{},
+	)
+}
+
+// Attrs are the config attributes for an RTSP camera model.
+type Attrs struct {
+	Address          string                             `json:"rtsp_address"`
+	IntrinsicParams  *transform.PinholeCameraIntrinsics `json:"intrinsic_parameters,omitempty"`
+	DistortionParams *transform.BrownConrady            `json:"distortion_parameters,omitempty"`
+}
+
+// Validate checks to see if the attributes of the model are valid.
+func (at *Attrs) Validate() error {
+	if prefix := strings.HasPrefix(at.Address, "rtsp://"); !prefix {
+		return errors.New(`rtsp_address must begin with "rtsp://"`)
+	}
+	if err := at.IntrinsicParams.CheckValid(); err != nil {
+		return err
+	}
+	if err := at.DistortionParams.CheckValid(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// rtspCamera contains the rtsp client, and the reader function that fulfills the camera interface.
+type rtspCamera struct {
+	gostream.VideoReader
+	u                       *url.URL
+	client                  *gortsplib.Client
+	cancelCtx               context.Context
+	cancelFunc              context.CancelFunc
+	activeBackgroundWorkers sync.WaitGroup
+	latestFrame             atomic.Value
+	logger                  golog.Logger
+}
+
+// Close closes the camera.
+func (rc *rtspCamera) Close(ctx context.Context) error {
+	var clientTerminated liberrors.ErrClientTerminated
+	rc.cancelFunc()
+	rc.activeBackgroundWorkers.Wait()
+	if err := rc.client.Close(); err != nil && !errors.Is(err, clientTerminated) {
```

Comment:
should closing the client having an error actually return an error or should it be logged instead?

Reply:
yea, just going to log instead.